src/viam/ur/module/ur_arm.cpp

```diff
@@ -172,128 +184,1284 @@ void write_waypoints_to_csv(const std::string& filepath, const std::list<Eigen::
+        if (!arm_conn_->driver->writeTrajectoryControlMessage(
+                urcl::control::TrajectoryControlMessage::TRAJECTORY_CANCEL, 0, RobotReceiveTimeout::off())) {
+            state.move_request_->cancel_error("failed to write trajectory control cancel message to URArm");
+            return event_local_mode_detected_{};
+        }
+    } else if (!state.move_request_->samples.empty() && state.move_request_->cancellation_request) {
+        // We have a move request that we haven't issued but a
+        // cancel is already pending. Don't issue it, just cancel it.
+        std::exchange(state.move_request_, {})->complete_cancelled();
+    } else {
+        // TODO: is it assured that we have positions/velocities here?
+        state.move_request_->write_joint_data(state.ephemeral_->joint_positions, state.ephemeral_->joint_velocities);
+    }
+
+    return std::nullopt;
+}
+
+URArm::state_::state_variant_ URArm::state_::state_controlled_::handle_event(event_connection_lost_) {
+    return state_disconnected_{};
+}
+
+URArm::state_::state_variant_ URArm::state_::state_controlled_::handle_event(event_estop_detected_) {
+    return state_independent_{std::move(arm_conn_), state_independent_::reason::k_estopped};
+}
+
+URArm::state_::state_variant_ URArm::state_::state_controlled_::handle_event(event_local_mode_detected_) {
+    return state_independent_{std::move(arm_conn_), state_independent_::reason::k_local_mode};
+}
+
+template <typename Event>
+URArm::state_::state_variant_ URArm::state_::state_controlled_::handle_event(Event) {
+    return std::move(*this);
+}
+
+URArm::state_::state_independent_::state_independent_(std::unique_ptr<arm_connection_> arm_conn, reason r)
+    : state_connected_(std::move(arm_conn)), reason_(r) {}
+
+std::string_view URArm::state_::state_independent_::name() {
+    using namespace std::literals::string_view_literals;
+    return "independent"sv;
+}
+
+std::string URArm::state_::state_independent_::describe() const {
+    std::ostringstream buffer;
+    buffer << name() << "(";
+    switch (reason_) {
+        case reason::k_estopped: {
+            buffer << "estop)";
+            break;
+        }
+        case reason::k_local_mode: {
+            buffer << "local)";
+            break;
+        }
+        case reason::k_both: {
+            buffer << "estop|local)";
+            break;
+        }
+        default: {
+            buffer << "unknown)";
+            break;
+        }
+    }
+    return buffer.str();
+}
+
+std::chrono::milliseconds URArm::state_::state_independent_::get_timeout() const {
+    // TODO: I'm not sure this actually makes sense: even if we are
+    // estopped, we still want high frequency service for the rtde
+    // interface.
+    if (estopped()) {
+        return k_estop_delay;
+    }
+    return k_noop_delay;
+}
+
+std::optional<URArm::state_::event_variant_> URArm::state_::state_independent_::upgrade_downgrade(state_&) {
```

Comment:
Expect changes here, @JohnN193 and I are still iterating on the exact recovery sequence for local mode.